### ISSUE TITLE
Add optional support for colourised output (where available...)

### ISSUE THREAD

**Issue by srcshelton** [github/super-linter#334]
**Is your feature request related to a problem? Please describe.**

When viewing `super-linter` output after running a local scan, it can be difficult to differentiate signal from the noise in the large number of lines generated.  Where available, showing errors in (say) red and warnings in (say) orange or yellow can vastly accelerate skim-reading the output.

**Describe the solution you'd like**

Many linters support colourising their output (when run on a supporting terminal) - it would be great if a configuration option were added (and potentially enabled by default for local runs?) which enabled coloured output for those linters which support it.

**Describe alternatives you've considered**

You could try to post-process the output to retrospectively add colour for specific linters... but honestly, that way madness lies!

**Comment by IAmHughes** [comment]
@srcshelton thanks for the suggestion! This is interesting, and I'll need to take a look at how the logging works with GitHub Actions. I haven't seen any way to colorize output, but that doesn't mean it's not possible. I'll do some research and report back tomorrow by end of day!

**Comment by srcshelton** [comment]
TBH, I was mainly considering running with `RUN_LOCAL=true`, and adding to the linter default options and available option to enable coloured output.

But if there is a sane means of getting GHA to give coloured output too, all the better!

(I've used some GH+CI/CD alternatives for work, and they do process ANSI escape sequences to provide coloured output when viewing pipeline logs - so there's a chance it's a relatively standard feature across web-based console output components?)

**Comment by IAmHughes** [comment]
> process ANSI escape sequences to provide coloured output when viewing pipeline logs 

That's intriguing, and I'll totally do some testing with that tomorrow as well 🎉.

**Comment by srcshelton** [comment]
Quick example: Running shellcheck from `koalaman/shellcheck:stable` automagically outputs bold and coloured results, whilst run from `super-linter` the output is plain.

**Comment by srcshelton** [comment]
(This is possibly mediated by the docker '`--tty`' option, which allows the running process to interrogate terminal capabilities.  Having said that, it may simply be a case of setting (or allowing the user to specify) a value for the environment variable `TERM` (to, for example, `vt100`)?

**Comment by IAmHughes** [comment]
@srcshelton at an initial glance, it looks like we DO support colorized output with ANSI tables (thanks for the suggestions!) so I'll be doing some testing of this tomorrow 🎉

**Comment by nemchik** [comment]
Colors in bash are pretty easily done once you dig into how to do it in a sane way.

https://github.com/GhostWriters/DockSTARTer/blob/5d8d79459a6976dc5aed3581cb780a238099b5ac/main.sh#L232-L316

This outputs colors where the terminal supports it. GHA does not support colors in any way that I've discovered (tried a lot), but this would work well for local runs. I'd be happy to make a PR if this method looks good to the code owners (doesn't have to use the logging functions I use, I just included that in the highlighting so you could see how I'm using the colors that get setup).

**Comment by nemchik** [comment]
https://github.com/github/super-linter/pull/346 will close this issue. Still working on it. Not too far off though.

**Comment by srcshelton** [comment]
> #346 will close this issue. Still working on it. Not too far off though.

This adds colour to `super-linter`'s direct output, but can coloured output also be enabled for linters which have options to add colour their individual outputs?

**Comment by nemchik** [comment]
That would be up to the linters to output the colors in ascii. Likely not much could be done in this repo to force that.

**Comment by srcshelton** [comment]
> That would be up to the linters to output the colors in ascii. Likely not much could be done in this repo to force that.

I'm not sure this is true - take `shellcheck`, for example: if the native binary is executed in a terminal, the output is colourised.  If their own docker container is run, this also outputs coloured text (when run from docker without specifically setting `TERM` or similar).  However, when `shellcheck` is run from `super-linter` the output is monochrome - so `super-linter` has either set something which inhibits coloured output, or hasn't set something required to get coloured output.

In the case of other linters, there may be command-line options which can be passed to force coloured output.

**Comment by nemchik** [comment]
Do you have an example of shellcheck output with color in GitHub actions using a method other than super-linter?

**Comment by srcshelton** [comment]
> Do you have an example of shellcheck output with color in GitHub actions using a method other than super-linter?

So far I've been discussing a `RUN_LOCAL=true` situation - if the GHA output formatter can't render ANSI escapes, then there's likely little can be done for the output under Actions unless this can be upgraded.

But even if GHA can't support coloured output, it would still be genuinely useful for local usage!

(And since local invocation is coloured *except* when using `super-linter`, this feels like a `super-linter` bug.  Or, at least, oversight)

**Comment by nemchik** [comment]
I get it now. Can you try with v3.2.0 just to confirm?

**Comment by IAmHughes** [comment]
GitHub Actions can support colored output with ANSI escapes. Here's one example I found: https://github.com/joshmgross/action-template/blob/main/src/utils.ts#L12

From some other documentation:
```
http://ascii-table.com/ansi-escape-sequences.php
https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit
Currently https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit  is NOT supported
We support sequences of format:
 Esc[CONTENTHEREm
 Where CONTENTHERE can be of format: VALUE;VALUE;VALUE or VALUE
     Where VALUE is SGR parameter https://www.vt100.net/docs/vt510-rm/SGR
         We support: 0 (reset), 1 (bold), 3 (italic), 4 (underline), 22 (not bold), 23 (not italic), 24 (not underline), 39 (default fg), 49 (default bg), 
                     fg colors - 30 (black), 31 (red), 32 (green), 33 (yellow), 34 (blue), 35 (magenta), 36 (cyan), 37 (white), 90 (grey)
                       with more brighness - 91 (red), 92 (green), 93 (yellow), 94 (blue), 95 (magenta), 96 (cyan), 97 (white)
                     bg colors - 40 (black), 41 (red), 42 (green), 43 (yellow), 44 (blue), 45 (magenta), 46 (cyan), 47 (white), 100 (grey)
                       with more brighness- 101 (red), 102 (green), 103 (yellow), 104 (blue), 105 (magenta), 106 (cyan), 107 (white)
 Where m refers to the "Graphics mode"
```

**Comment by srcshelton** [comment]
> I get it now. Can you try with v3.2.0 just to confirm?

Sure thing!  Give me a tick...

**Comment by srcshelton** [comment]
Hmm - might have spotted a problem here :(

```
$ docker run --cap-drop ALL --init --rm --volume /root/docker-build:/mnt:ro koalaman/shellcheck:stable --color=always /mnt/lint-test.sh

In /mnt/lint-test.sh line 13:
red=''
^-^ SC2034: red appears unused. Verify use (or export if used externally).


In /mnt/lint-test.sh line 78:
                . web/webapps.conf >/dev/null 2>&1
                  ^--------------^ SC1091: Not following: web/webapps.conf was not specified as input (see shellcheck -x).


In /mnt/lint-test.sh line 167:
                  ${fs_name:+--build-arg service_fs_name="${fs_name}"} \
                                                          ^--------^ SC2154: fs_name is referenced but not assigned.


In /mnt/lint-test.sh line 189:
cat file | grep thing && echo yes || echo no
    ^--^ SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- red appears unused. Verify use (o...
  https://www.shellcheck.net/wiki/SC2154 -- fs_name is referenced but not ass...
  https://www.shellcheck.net/wiki/SC1091 -- Not following: web/webapps.conf w...
```

... vs:

```
$ docker run --env DEFAULT_WORKSPACE=/lint --env RUN_LOCAL=true --env TERM=screen-256color --cap-drop ALL --init --interactive --rm --tty --volume /root/docker-build/lint-test.sh:/lint:ro github/super-linter:v3.2.0
--------------------------------------------------------------------------------

                              /@@#///////@@/(@//@%/(@.@(       @@
                          @@//////////////////////////////#*  @@@
                        @////@//(///////////@@@@@///@//@/@**//@@(
                      @///////@///////////////@@@@    (           @,
                     @/(&/@////////////////////                     @
                    @////////////////////////@@                      @
                  @%////////(//////////%/////&@            @@       *,@
             @@@@@/@/#/////(&//////////////////                       .@
        *@@@@@.    .%///(//@//////////////////&.   .@@,                 @%
      @@%           .&@&&/@.@//&/////(//////////    @@@@@@@@@         .. &@
    @@%               @@@@@   @&/////////////////#   @/       V  @@/ ,@@@ @
@@@%                   @@@@        .%@@@@//////#@ @   @@         @     .,.
                                          @@@/@(  (@@@@% @/\      %
                                           @@@@(    .     .@@/\   #
                                             @                  %@%

--------------------------------------------------------------------------------

---------------------------------------------
--- GitHub Actions Multi Language Linter ----
---------------------------------------------

---------------------------------------------
The Super-Linter source code can be found at:
 - https://github.com/github/super-linter
---------------------------------------------
--------------------------------------------
Gathering GitHub information...
NOTE: ENV VAR [RUN_LOCAL] has been set to:[true]
bypassing GitHub Actions variables...
Linting all files in mapped directory:[/lint]

--------------------------------------------
Gathering user validation information...

----------------------------------------------
----------------------------------------------
The script has completed
----------------------------------------------
----------------------------------------------


All file(s) linted successfully with no errors detected
----------------------------------------------
```

... although the `All file(s) linted successfully with no errors detected` *does* appear in green.

I may have misunderstood the use of `DEFAULT_WORKSPACE` - trying again now...

**Comment by nemchik** [comment]
I maybe know where the issue is. I'll have a look tomorrow.

**Comment by admiralAwkbar** [comment]
This seems to have been completed and shipped...

**Comment by nemchik** [comment]
> This seems to have been completed and shipped...

Code in this repo outputs colors, but the linters should also be able to output colors (locally for sure, but also possibly in CI).

I'll still do some investigating today to see if I can narrow down getting that last part done.

**Comment by nemchik** [comment]
Ok I've finally got some conclusive results.

* I started by making a script named `main.sh` in my current directory and intentionally put an error that shellcheck would flag in the file.
* Next I ran `docker run --rm -it -e RUN_LOCAL=true --entrypoint /bin/bash -v ${PWD}/main.sh:/tmp/lint/main.sh github/super-linter` which brought me into the container interactively.
* I ran `/action/lib/linter.sh` which did not give me any colorized output except the last line `Exiting with errors found!` in red. More of the output should have been in color from the code in this repo that should now output color.
* I created a file `try.sh` in the same directory as `main.sh` and just put in a shebang and a line that had `shellcheck main.sh`. this output color correctly.
* I modified `try.sh` to run `MYVAR="$(shellcheck main.sh)"` then `echo -e "$MYVAR"` (which is quite similar to how this repo operates, but overly simplified). NO color would output.
* I did some digging and found that shellcheck has a `--color` option similar to `ls` so I changed `try.sh` to `MYVAR="$(shellcheck --color main.sh)"` and this made the output show in color!
* After all that I am still having some trouble with the output of the code in this repo not being properly colorized where I would expect it to be (see screenshot)

![image](https://user-images.githubusercontent.com/725456/86403978-5bc3ac80-bc74-11ea-8357-ddeccb0b3652.png)
The areas highlighted in the red boxes should be colored.

Anyway what this all tells me is that each individual linter may or may not by default output colorized text and some may need the color option enabled. I tested enabling color with the code in this repo for shellcheck and will be able to make a PR for that. For the rest we will have to determine if a color option exists.

I think for now we can close this issue and potentially address each linter tool used in this repo individually with a separate issue or PR.
